runtime/lua/vim/lsp/rpc.lua

```diff
@@ -139,17 +139,26 @@ local function request_parser_loop()
       -- TODO(ashkan) I'd like to remove this, but it seems permanent :(
       local buffer_start = buffer:find(header_start_pattern)
       local headers = parse_headers(buffer:sub(buffer_start, start-1))
-      buffer = buffer:sub(finish+1)
       local content_length = headers.content_length
+      local body_chunks = {buffer:sub(finish+1)}
+      local body_length = #body_chunks[1]
       -- Keep waiting for data until we have enough.
-      while #buffer < content_length do
-        buffer = buffer..(coroutine.yield()
-            or error("Expected more data for the body. The server may have died.")) -- TODO hmm.
+      while body_length < content_length do
+        local chunk = coroutine.yield()
+            or error("Expected more data for the body. The server may have died.") -- TODO hmm.
+        table.insert(body_chunks, chunk)
+        body_length = body_length + #chunk
       end
-      local body = buffer:sub(1, content_length)
-      buffer = buffer:sub(content_length + 1)
+      local last_chunk = body_chunks[#body_chunks]
+
+      body_chunks[#body_chunks] = last_chunk:sub(1, content_length - body_length - 1)
+      local rest = ''
+      if body_length > content_length then
+        rest = last_chunk:sub(content_length - body_length)
+      end
+      local body = table.concat(body_chunks)
       -- Yield our data.
-      buffer = buffer..(coroutine.yield(headers, body)
+      buffer = rest..(coroutine.yield(headers, body)
           or error("Expected more data for the body. The server may have died.")) -- TODO hmm.
     else
       -- Get more data since we don't have enough.
```

Comment:
I'm guessing you're not modifying the repeat concatenation below because the header is guaranteed to be small? (I'm wondering whether this case every gets triggered, as a header should be <100 bytes).

Reply:
One reason is that the header will be small enough. Another is, I thought finding the border between header part and body part(actually double CRLFs) will be too complicated task for string chunks table.(currently [this line](https://github.com/neovim/neovim/blob/09c237c13c20f4df7b1a06f353574e910408da46/runtime/lua/vim/lsp/rpc.lua#L131))